File: 55. Jump Game/55.go
package leetcode

func canJump(nums []int) bool {
	numsLen := len(nums)-1
	startIndex := 0
	endIndex := nums[0]
	for{
		if endIndex>=numsLen{
			return true
		}
		bigvalue := -1
		for i:=startIndex; i<endIndex+1; i++{
			tmp := nums[i] + i
			if tmp>numsLen{
				return true
			}
			if tmp>bigvalue{
				bigvalue = tmp
			}
		}
		startIndex = endIndex
		endIndex = bigvalue
		if endIndex<=startIndex{
			return false
		}
	}
}

func canJump1(nums []int) bool{
	numsLen := len(nums)-1
	for i:=numsLen; i>=0; i--{
		if nums[i] + i >= numsLen{
			numsLen = i
		}
	}
	return numsLen == 0
}

File: 55. Jump Game/_55.py
class Solution:
    def canJump(self, nums) -> bool:
        lenNums = len(nums)-1
        stratindex = 0
        endindex = nums[0]
        while True:
            if endindex >= lenNums:
                return True
            bigvalue = -1
            for i in range(stratindex, endindex+1):
                tmp = nums[i]+i
                if tmp >= lenNums:
                    return True
                if tmp > bigvalue:
                    bigvalue = tmp
            stratindex = endindex
            endindex = bigvalue
            if stratindex>=bigvalue:
                return False

    # 贪婪算法
    def canJump1(self, nums) -> bool:
        lenNums = len(nums)-1
        for i in range(lenNums, -1, -1):
            if nums[i]+i >= lenNums:
                lenNums = i
        return lenNums == 0